[Fit-API] Adress PR comments

roywei opened the pull request:
## Description ##
To address PR comments in https://github.com/apache/incubator-mxnet/pull/14629
1. Check train_data and val_data only supports dataloader, give proper error msg
2. exposed batch_axis and set default 0
3.  updated initialization logic, so fine-tuning use case (net partially initialized) will not get forced re-init.
4. explained auto mode in checkpoint and early stopping, added docstring, warning and logging message when auto mode is used and np.greater/np.less is determined in auto mode.
5. log learning rate every epoch
6. update logic to save checkpoint based on epoch_period(default 1) and batch_period (default None)
7. update checkpoint handler to save the following: symbol (1 time after first batch, if net is hybridizable), params and states per epcoh or batch, only keep a maximum number of checkpoints(default 5)
8. added feature to resume training from checkpoints

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

anirudhacharya commented:
@mxnet-label-bot add [pr-awaiting-review]

eric-haibin-lin commented on python/mxnet/gluon/contrib/estimator/estimator.py:
```diff
@@ -190,26 +195,29 @@ def prepare_loss_and_metrics(self):
          val_data : DataLoader
-             validation data with data and labels
+             validation data loader with data and labels
+         batch_axis : int, default 0
```
Better to reorder the doc so that batch_axis appears last?

eric-haibin-lin commented on python/mxnet/gluon/contrib/estimator/event_handler.py:
```diff
@@ -263,108 +268,171 @@ def epoch_end(self, estimator, *args, **kwargs):
 
-    :py:class:`CheckpointHandler` save the network parameters every epoch
+    :py:class:`CheckpointHandler` saves the network architecture after first batch if the model
+    can be hybridized(), saves model parameters and trainer states after user defined period,
```
can be hybridized() -> can be fully hybridized. 
`hybridized()` is not a function

eric-haibin-lin commented on python/mxnet/gluon/contrib/estimator/event_handler.py:
```diff
@@ -263,108 +268,171 @@ def epoch_end(self, estimator, *args, **kwargs):
+        epoch intervals between saving the network
+    batch_period: int, default None
+        batch intervals between saving the network,
+        by default don't save any checkpoint based on number of batches
```
by default, checkpoints are not saved based on the number of batches

eric-haibin-lin commented on python/mxnet/gluon/contrib/estimator/event_handler.py:
```diff
@@ -263,108 +268,171 @@ def epoch_end(self, estimator, *args, **kwargs):
                 if np.isnan(monitor_value):
-                    warnings.warn(RuntimeWarning('%s is not updated, make sure you pass one of the metric objects'
-                                                 'as monitor, you can use estimator.prepare_loss_and_metrics to'
+                    warnings.warn(RuntimeWarning('SKipping save best because %s is not updated, make sure you '
```
SKipping -> Skipping

eric-haibin-lin commented on python/mxnet/gluon/contrib/estimator/estimator.py:
```diff
@@ -222,28 +230,30 @@ def evaluate(self,
     def fit(self, train_data,
```
Any thought on how to make resuming from the checkpoint easier? With the current code, if I want to train 90 epochs and the training failed at epoch 30, the steps I need to do to resume training are:
- create net
- call net.load_parameters, with the correct .params filename
- create trainer
- call trainer.load_states, with the correct .states filename
- create an estimator
- call fit with epoch = 60? 

Since we know the naming scheme of the checkpoint, we can do all these automatically by loading the last checkpoint?

roywei commented:
@eric-haibin-lin  addressed comments, added option to resume from checkpoint. Will load params and trainer states at train begin if checkpoint files detected. Will train the remain of the epochs.

eric-haibin-lin commented:
LGTM. My comments are addressed.
